src/conf/domain_conf.c

```diff
@@ -19897,19 +19901,69 @@ virDomainDefParseXML(xmlXPathContextPtr ctxt,
                                                        ctxt,
                                                        nodes[i],
                                                        flags);
+
         if (!chr)
             return NULL;
 
-        if (chr->target.port == -1) {
+        def->serials[def->nserials++] = chr;
+
+        // Giving precedence to the isa-serial device since
+        // only limited ports can be used for such devices.
+        if (chr->targetType == VIR_DOMAIN_CHR_SERIAL_TARGET_MODEL_ISA_SERIAL) {
+            // Taking the isa serial deivces to start of the array.
+            for (j = def->nserials; j > isa_serial_count; j--)
+                def->serials[j] = def->serials[j-1];
+            def->serials[isa_serial_count++] = chr;
+        }
+
+        // Maintaining the buffer for first max_available_isa_serial_ports unused ports.
+        if (chr->target.port != -1 && chr->target.port <= max_available_isa_serial_ports) {
+            if (used_serial_port_buffer & (1<<chr->target.port)) {
+                virReportError(VIR_ERR_INTERNAL_ERROR,
+                    _("target port [%d] already allocated."),
+                    chr->target.port);
+                return NULL;
+            }
+            used_serial_port_buffer |= 1<<chr->target.port;
+        }
+    }
+
+    // Assign the ports to the devices.
+    for (i = 0; i < n; i++) {
+        if (def->serials[i]->target.port != -1) continue;
```

Comment:
should you check for conflict for non-isa too.

Reply:
Yeah doing that already. hence using max_serial_port.

Reply:
yes but max_serial_port should come only if user did not provide ports.  What if user passes duplicate ports. Should you give error. But i think it should be fine as those are not even used.

Reply:
already doing in 19920